File: Singleton_Pattern/go/Singleton_Pattern.go
package DesginPattern

import (
	"sync"
)

type Singleton struct{}

var instance *Singleton
var hungryinstance *Singleton
var mu sync.Mutex

//double check lock for lazy
func GetInstance() *Singleton {
	if instance == nil {
		mu.Lock()
		defer mu.Unlock()
		if instance == nil {
			instance = &Singleton{}
		}
	}
	return instance
}
func GetInstanceHungry() *Singleton {
	return hungryinstance
}

//hungry model
func init() {
	hungryinstance = &Singleton{}
}

File: Singleton_Pattern/cpp/Singleton_Pattern.cc
#include<iostream>
#include<mutex>
using namespace std;


std::mutex mtx;
class Singleton {
public:
    static Singleton* Instance;
    static Singleton* GetInstance() {
        if (Instance == NULL) {
            mtx.lock();
            if (Instance == NULL) {
                Instance = new Singleton;
            }
            mtx.unlock();
        }
        return Instance;
    }
private:
    Singleton(){};
public:
    Singleton(const Singleton&) = delete;
};

Singleton* Singleton::Instance = NULL;


int main() {
    auto a = Singleton::GetInstance();
    return 0;
}